keps/sig-network/2004-topology-aware-routing/README.md

```diff
@@ -0,0 +1,600 @@
+  - Balanced: Use the original approach to ensure traffic is distributed across
+  all endpoints for a Service.
+  - RequireZone: Only route to endpoints within the same zone.
+  - PreferZone: Prefer routing to endpoints within the same zone when they are
+  available.
+- Use the standardized topology label `topology.kubernetes.io/zone` to derive
+the zones of nodes and endpoints.
+- Use EndpointSlice subsetting as the primary mechanism for topology aware
+routing.
+  - Improve kube-proxy scalability by limiting the number of endpoints it needs
+  to process.
+- Minimize churn of EndpointSlices while doing topology aware distribution.
+- Minimize the number of new EndpointSlices required.
+- Provide a simple API that does not require any new API fields.
+
+### Non-Goals
+- Real-time distribution rebalancing based on traffic load or distribution
+feedback or metrics.
+- Providing a way for kube-proxy to consume subsets of EndpointSlices (this is
+a separate KEP).
+- Providing a way to require Node local traffic ([this is a separate KEP](https://github.com/kubernetes/enhancements/pull/1944)).
+- Multi-cluster topology aware routing (this same pattern may be useful there
+though).
+- Region based topology aware routing (this may come later).
+- Ensuring that endpoints are distributed evenly across zones.
+
+## Proposal
+
+### Terminology
+
+- **EndpointSlice Group**: One or multiple EndpointSlices that will be consumed
+by the same set of zones.
+- **Endpoints**: In this KEP, “endpoints” are used to describe individual Pod
+IPs. These are not references to Kubernetes Endpoints resources. We also omit
+the unready pods that number of endpoints = number of pods.
+
+### Configuration
+
+This proposal builds off of the [EndpointSlice API](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/).
+
+We propose a new Service annotation: `service.kubernetes.io/same-zone`
+with three options:
+- `Balanced`: This is the default, it does not need to be specified which also
+provides backward compatibility for users. It represents the existing approach
+that is already used, no special mechanism will be added. Traffic will be
+evenly distributed among endpoints behind a Service.
+- `Require`: EndpointSlice controller will conduct a simple but aggressive
+routing mechanism. It will group endpoints in a zone to an EndpointSlice Group
+that only receives traffic from that zone. `Require` requires users to ensure
+that a sufficient number of endpoints for a Service exist in each zone.
+Services using this approach will only be reachable from zones that contain
+endpoints for the Service.
+- `Prefer`: EndpointSlice controller will conduct a more conservative routing
+mechanism. It will keep traffic in zone as long as each zone has a sufficient
+number of endpoints. If a zone does not have a sufficient number of endpoints,
+traffic will be routed to other zones.
+
+This same pattern could be used for topology aware routing at node and region
```

Comment:
I think Zone and Region are very closely related. Node is definitely more of a stretch. This becomes a bit of a question of what our config is saying though. If the config is saying this Service should prefer/require routing to the same node/zone/region, I think it makes sense to keep the config consistent (as currently proposed). On the other hand, if we're more focused on this feature as "how we subset EndpointSlices" and the configuration reflects that, I don't think we should have the same config for node-local traffic.

Reply:
I am leaning pretty hard away from multiple knobs here.  I think something more abstract makes more sense.  Especially if you think about multi-cluster Services here.  E.g. If the MCS thingy already does some form of subsetting, what does this policy mean?  @JeremyOT

Reply:
I explored combining this into a single setting and it got pretty overwhelming pretty quickly, especially if nodes were involved. Here's a [doc with some of the potential combinations this would result in](https://docs.google.com/document/d/1Iul7rRTx_knezvPdwqZ9SQxEYUUXMTDnjCeYmNmcEGA/edit?usp=sharing). Even if we were just covering zone and region, we'd potentially need to support 6 different values: 

- RequireZone
- PreferZone
- PreferZonePreferRegion
- PreferZoneRequireRegion
- RequireRegion
- PreferRegion

If we want to support different settings for zone and region in the future, I think it's preferable to split that config out instead of squashing everything into a single field.

Reply:
Since there's a lot of comments/threads here, I'll copy some relevant points from another thread:

We could transition to a single EndpointSlice specific annotation/field like `endpointslice.kubernetes.io/slicing-policy` and support the following values:

- One of `None`, `Random`, `Shared`, or `Balanced`
- `PreferZone`
- `RequireZone`

Reply:
As a follow up here, we could potentially simplify the above options into:

- RequireZone
- PreferZone
- RequireRegion (infers PreferZone)
- PreferRegion (infers PreferZone)

Reply:
I'm a little worried that `slicing-policy` is too close to the actual implementation details here, but that does make room for more advanced topology aware routing solutions to be built on top of this in the future.